flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -369,5 +420,46 @@ public ProducerRecord<byte[], byte[]> serialize(
                     value,
                     headerProvider != null ? headerProvider.getHeaders(element) : null);
         }
+
+        @Override
+        public Optional<KafkaDatasetFacet> getKafkaDatasetFacet() {
+            if (!(topicSelector instanceof KafkaDatasetIdentifierProvider)) {
+                LOG.info("Cannot identify topics. Not an TopicsIdentifierProvider");
+                return Optional.empty();
+            }
+
+            Optional<DefaultKafkaDatasetIdentifier> topicsIdentifier =
+                    ((KafkaDatasetIdentifierProvider) (topicSelector)).getDatasetIdentifier();
+
+            if (!topicsIdentifier.isPresent()) {
+                LOG.info("No topics' identifiers provided");
+                return Optional.empty();
+            }
+
+            return Optional.of(new DefaultKafkaDatasetFacet(topicsIdentifier.get()));
+        }
+
+        @Override
+        public Optional<TypeDatasetFacet> getTypeDatasetFacet() {
+            if (this.valueSerializationSchema instanceof ResultTypeQueryable) {
+                return Optional.of(
+                        new DefaultTypeDatasetFacet(
+                                ((ResultTypeQueryable<?>) this.valueSerializationSchema)
+                                        .getProducedType()));
+            } else {
+                // gets type information from serialize method signature
+                TypeToken serializationSchemaType =
+                        TypeToken.of(valueSerializationSchema.getClass());
+                Class parameterType =
+                        serializationSchemaType
```

Comment:
Is there a way to avoid using this reflection - instanceof and Class. Maybe using a config driven approach and java SPI. The connectors / formats bring in serialization implementations in this way that avoid the overhead of reflection.

Reply:
You can avoid this with implementing `ResultTypeQueryable` of the value serialisation schema.